Not all individual audit files are being exported

#### Software versions 
Briefcase v1.11.x

#### Problem description
When exporting forms that have audit data, due to concurrency and delays while writing to the filesystem, Briefcase can fail to write some of the individual audit files to the `media` export directory.

#### Steps to reproduce the problem
Unzip the contents of [this zip](https://github.com/opendatakit/briefcase/files/2486693/default.zip) into Briefcase's storage directory and try to export the form inside.

#### Expected behavior
We should get all the individual audit files without errors.

We should use the submission's instance ID as a suffix for the individual audit filenames instead of a sequence number.

#### Other information 
This is a stack trace of the error:

```
2018-10-17 11:25:48,704 [Thread-3] ERROR o.o.briefcase.ui.export.ExportPanel - Error while exporting forms
java.io.UncheckedIOException: java.nio.file.FileAlreadyExistsException: /tmp/media/audit.csv
	at org.opendatakit.briefcase.reused.UncheckedFiles.copy(UncheckedFiles.java:114)
	at org.opendatakit.briefcase.export.CsvFieldMappers.binary(CsvFieldMappers.java:201)
	at org.opendatakit.briefcase.export.CsvFieldMappers.lambda$null$0(CsvFieldMappers.java:64)
	at java.util.Optional.map(Optional.java:215)
	at org.opendatakit.briefcase.export.CsvFieldMappers.lambda$static$1(CsvFieldMappers.java:64)
	at org.opendatakit.briefcase.export.CsvFieldMapper.lambda$compose$0(CsvFieldMapper.java:65)
	at org.opendatakit.briefcase.export.CsvFieldMapper.lambda$decorateOutput$1(CsvFieldMapper.java:76)
	at org.opendatakit.briefcase.export.CsvFieldMappers.lambda$nonRepeatableGroup$16(CsvFieldMappers.java:260)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at org.opendatakit.briefcase.export.CsvSubmissionMappers.lambda$main$2(CsvSubmissionMappers.java:64)
	at org.opendatakit.briefcase.export.ExportToCsv.lambda$null$5(ExportToCsv.java:124)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
	at java.util.stream.ReferencePipeline$11$1.accept(ReferencePipeline.java:373)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
	at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.nio.file.FileAlreadyExistsException: /tmp/media/audit.csv
	at sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:551)
	at sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:253)
	at java.nio.file.Files.copy(Files.java:1274)
	at org.opendatakit.briefcase.reused.UncheckedFiles.copy(UncheckedFiles.java:112)
	... 47 common frames omitted
```

--- a/src/org/opendatakit/briefcase/export/CsvFieldMappers.java
+++ b/src/org/opendatakit/briefcase/export/CsvFieldMappers.java
@@ -66,7 +66,7 @@ final class CsvFieldMappers {
 
   private static CsvFieldMapper AUDIT_MAPPER = BINARY_MAPPER
       .andThen((formName, localId, workingDir, model, maybeElement, configuration) -> maybeElement
-          .map(e -> audit(formName, localId, workingDir, configuration, e))
+          .map(e -> aggregatedAuditFile(formName, localId, workingDir, configuration, e))
           .orElse(empty(model.fqn())))
       .map(output -> output.filter(pair -> !pair.getLeft().contains("-aggregated")));
 
@@ -227,7 +227,7 @@ final class CsvFieldMappers {
     return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(sequentialDestinationFile.getFileName()).toString()));
   }
 
-  private static Stream<Pair<String, String>> audit(String formName, String localId, Path workingDir, ExportConfiguration configuration, XmlElement e) {
+  private static Stream<Pair<String, String>> aggregatedAuditFile(String formName, String localId, Path workingDir, ExportConfiguration configuration, XmlElement e) {
     if (!e.hasValue())
       return empty(e.fqn() + "-aggregated");
 

--- a/src/org/opendatakit/briefcase/export/CsvFieldMappers.java
+++ b/src/org/opendatakit/briefcase/export/CsvFieldMappers.java
@@ -15,6 +15,7 @@
  */
 package org.opendatakit.briefcase.export;
 
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
 import static java.nio.file.StandardOpenOption.APPEND;
 import static java.text.DateFormat.getDateInstance;
 import static java.text.DateFormat.getDateTimeInstance;
@@ -34,6 +35,7 @@ import static org.opendatakit.briefcase.reused.UncheckedFiles.getMd5Hash;
 import static org.opendatakit.briefcase.reused.UncheckedFiles.lines;
 import static org.opendatakit.briefcase.reused.UncheckedFiles.stripFileExtension;
 import static org.opendatakit.briefcase.reused.UncheckedFiles.write;
+import static org.opendatakit.briefcase.util.StringUtils.stripIllegalChars;
 import static org.opendatakit.common.utils.WebUtils.parseDate;
 
 import java.nio.file.Files;
@@ -60,8 +62,8 @@ import org.opendatakit.briefcase.reused.Pair;
 final class CsvFieldMappers {
   private static final Map<DataType, CsvFieldMapper> mappers = new HashMap<>();
 
-  private static CsvFieldMapper INDIVIDUAL_FILE_AUDIT_MAPPER = (__, ___, workingDir, field, element, configuration) -> element
-      .map(e -> individualAuditFile(e, workingDir, configuration))
+  private static CsvFieldMapper INDIVIDUAL_FILE_AUDIT_MAPPER = (__, instanceId, workingDir, field, maybeElement, configuration) -> maybeElement
+      .map(element -> individualAuditFile(instanceId, workingDir, configuration, element))
       .orElse(empty(field.fqn()));
 
   private static CsvFieldMapper AGGREGATED_FILE_AUDIT_MAPPER = (formName, localId, workingDir, model, maybeElement, configuration) -> maybeElement
@@ -231,7 +233,7 @@ final class CsvFieldMappers {
     return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(sequentialDestinationFile.getFileName()).toString()));
   }
 
-  private static Stream<Pair<String, String>> individualAuditFile(XmlElement element, Path workingDir, ExportConfiguration configuration) {
+  private static Stream<Pair<String, String>> individualAuditFile(String instanceId, Path workingDir, ExportConfiguration configuration, XmlElement element) {
     // TODO We should separate the side effect of writing files to disk from the csv output generation
 
     if (!element.hasValue())
@@ -247,41 +249,12 @@ final class CsvFieldMappers {
 
     Path sourceFile = workingDir.resolve(sourceFilename);
 
-    // When the source file doesn't exist, we return the input value
     if (!exists(sourceFile))
       return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(sourceFilename).toString()));
 
-    // When the destination file doesn't exist, we copy the source file
-    // there and return its path relative to the instance folder
-    Path destinationFile = configuration.getExportMediaPath().resolve(sourceFilename);
-    if (!exists(destinationFile)) {
-      copy(sourceFile, destinationFile);
-      return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(destinationFile.getFileName()).toString()));
-    }
-
-    // When the destination file has the same hash as the source file,
-    // we don't do any side-effect and return its path relative to the
-    // instance folder
-    Boolean sameHash = OptionalProduct.all(getMd5Hash(sourceFile), getMd5Hash(destinationFile)).map(Objects::equals).orElse(false);
-    if (sameHash)
-      return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(destinationFile.getFileName()).toString()));
-
-    // When the hashes are different, we compute the next sequential suffix for
-    // the a new destination file to avoid overwriting the one we found already
-    // there. We try every number in the sequence until we find one that won't
-    // produce a destination file that exists in the output directory.
-    String namePart = stripFileExtension(sourceFilename);
-    String extPart = getFileExtension(sourceFilename).map(extension -> "." + extension).orElse("");
-    int sequenceSuffix = 2;
-    Path sequentialDestinationFile;
-    do {
-      sequentialDestinationFile = configuration.getExportMediaPath().resolve(String.format("%s-%d%s", namePart, sequenceSuffix++, extPart));
-    } while (exists(sequentialDestinationFile));
-
-    // Now that we have a valid destination file, we copy the source file
-    // there and return its path relative to the instance folder
-    copy(sourceFile, sequentialDestinationFile);
-    return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(sequentialDestinationFile.getFileName()).toString()));
+    Path destinationFile = configuration.getExportMediaPath().resolve("audit-" + stripIllegalChars(instanceId) + ".csv");
+    copy(sourceFile, destinationFile, REPLACE_EXISTING);
+    return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(destinationFile.getFileName()).toString()));
   }
 
   private static Stream<Pair<String, String>> aggregatedAuditFile(String formName, String localId, Path workingDir, ExportConfiguration configuration, XmlElement e) {

--- a/src/org/opendatakit/briefcase/export/CsvFieldMappers.java
+++ b/src/org/opendatakit/briefcase/export/CsvFieldMappers.java
@@ -60,14 +60,16 @@ import org.opendatakit.briefcase.reused.Pair;
 final class CsvFieldMappers {
   private static final Map<DataType, CsvFieldMapper> mappers = new HashMap<>();
 
-  private static final CsvFieldMapper BINARY_MAPPER = (__, ___, workingDir, field, element, configuration) -> element
-      .map(e -> binary(e, workingDir, configuration))
+  private static CsvFieldMapper INDIVIDUAL_FILE_AUDIT_MAPPER = (__, ___, workingDir, field, element, configuration) -> element
+      .map(e -> individualAuditFile(e, workingDir, configuration))
       .orElse(empty(field.fqn()));
 
-  private static CsvFieldMapper AUDIT_MAPPER = BINARY_MAPPER
-      .andThen((formName, localId, workingDir, model, maybeElement, configuration) -> maybeElement
-          .map(e -> aggregatedAuditFile(formName, localId, workingDir, configuration, e))
-          .orElse(empty(model.fqn())))
+  private static CsvFieldMapper AGGREGATED_FILE_AUDIT_MAPPER = (formName, localId, workingDir, model, maybeElement, configuration) -> maybeElement
+      .map(e -> aggregatedAuditFile(formName, localId, workingDir, configuration, e))
+      .orElse(empty(model.fqn()));
+
+  private static CsvFieldMapper AUDIT_MAPPER = INDIVIDUAL_FILE_AUDIT_MAPPER
+      .andThen(AGGREGATED_FILE_AUDIT_MAPPER)
       .map(output -> output.filter(pair -> !pair.getLeft().contains("-aggregated")));
 
 
@@ -82,7 +84,9 @@ final class CsvFieldMappers {
     mappers.put(GEOPOINT, simpleMapper(CsvFieldMappers::geopoint, 4));
 
     // Binary fields require knowledge of the export configuration and working dir
-    mappers.put(BINARY, BINARY_MAPPER);
+    mappers.put(BINARY, (__, ___, workingDir, field, maybeElement, configuration) -> maybeElement
+        .map(element -> binary(element, workingDir, configuration))
+        .orElse(empty(field.fqn())));
 
     // Null fields encode groups (repeating and non-repeating), therefore,
     // they require the full context
@@ -175,6 +179,59 @@ final class CsvFieldMappers {
   }
 
   private static Stream<Pair<String, String>> binary(XmlElement element, Path workingDir, ExportConfiguration configuration) {
+    // TODO We should separate the side effect of writing files to disk from the csv output generation
+
+    if (!element.hasValue())
+      return empty(element.fqn());
+
+    String sourceFilename = element.getValue();
+
+    if (!configuration.resolveExportMedia())
+      return Stream.of(Pair.of(element.fqn(), sourceFilename));
+
+    if (!Files.exists(configuration.getExportMediaPath()))
+      createDirectories(configuration.getExportMediaPath());
+
+    Path sourceFile = workingDir.resolve(sourceFilename);
+
+    // When the source file doesn't exist, we return the input value
+    if (!exists(sourceFile))
+      return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(sourceFilename).toString()));
+
+    // When the destination file doesn't exist, we copy the source file
+    // there and return its path relative to the instance folder
+    Path destinationFile = configuration.getExportMediaPath().resolve(sourceFilename);
+    if (!exists(destinationFile)) {
+      copy(sourceFile, destinationFile);
+      return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(destinationFile.getFileName()).toString()));
+    }
+
+    // When the destination file has the same hash as the source file,
+    // we don't do any side-effect and return its path relative to the
+    // instance folder
+    Boolean sameHash = OptionalProduct.all(getMd5Hash(sourceFile), getMd5Hash(destinationFile)).map(Objects::equals).orElse(false);
+    if (sameHash)
+      return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(destinationFile.getFileName()).toString()));
+
+    // When the hashes are different, we compute the next sequential suffix for
+    // the a new destination file to avoid overwriting the one we found already
+    // there. We try every number in the sequence until we find one that won't
+    // produce a destination file that exists in the output directory.
+    String namePart = stripFileExtension(sourceFilename);
+    String extPart = getFileExtension(sourceFilename).map(extension -> "." + extension).orElse("");
+    int sequenceSuffix = 2;
+    Path sequentialDestinationFile;
+    do {
+      sequentialDestinationFile = configuration.getExportMediaPath().resolve(String.format("%s-%d%s", namePart, sequenceSuffix++, extPart));
+    } while (exists(sequentialDestinationFile));
+
+    // Now that we have a valid destination file, we copy the source file
+    // there and return its path relative to the instance folder
+    copy(sourceFile, sequentialDestinationFile);
+    return Stream.of(Pair.of(element.fqn(), Paths.get("media").resolve(sequentialDestinationFile.getFileName()).toString()));
+  }
+
+  private static Stream<Pair<String, String>> individualAuditFile(XmlElement element, Path workingDir, ExportConfiguration configuration) {
     // TODO We should separate the side effect of writing files to disk from the csv output generation
 
     if (!element.hasValue())

--- a/test/java/org/opendatakit/briefcase/export/CsvFieldMappersTest.java
+++ b/test/java/org/opendatakit/briefcase/export/CsvFieldMappersTest.java
@@ -267,6 +267,20 @@ public class CsvFieldMappersTest {
   }
 
   @Test
+  public void individual_audit_filenames_include_the_instance_id() {
+    scenario = Scenario.nonGroup("some-form", DataType.BINARY, "audit", "meta");
+    write(scenario.getWorkDir().resolve("audit.csv"), "event, node, start, end\nform start,,1536663986578,\n");
+
+    List<Pair<String, String>> output = scenario.mapSimpleValue("audit.csv", true);
+    assertThat(output, contains(Pair.of("data-audit", "media/audit-" + scenario.getInstanceId() + ".csv")));
+
+    Path outputAudit = scenario.getOutputMediaDir().resolve("audit-" + scenario.getInstanceId() + ".csv");
+
+    assertThat(outputAudit, exists());
+    assertThat(outputAudit, fileContains("event, node, start, end\nform start,,1536663986578,\n"));
+  }
+
+  @Test
   public void audit_fields_append_the_submissions_content_to_the_output_audit_file() {
     scenario = Scenario.nonGroup("some-form", DataType.BINARY, "audit", "meta");
     write(scenario.getWorkDir().resolve("audit.csv"), "event, node, start, end\nform start,,1536663986578,\n");

--- a/test/java/org/opendatakit/briefcase/export/CsvFieldMappersTest.java
+++ b/test/java/org/opendatakit/briefcase/export/CsvFieldMappersTest.java
@@ -285,10 +285,7 @@ public class CsvFieldMappersTest {
     scenario = Scenario.nonGroup("some-form", DataType.BINARY, "audit", "meta");
     write(scenario.getWorkDir().resolve("audit.csv"), "event, node, start, end\nform start,,1536663986578,\n");
 
-    List<Pair<String, String>> output = scenario.mapSimpleValue("audit.csv", true);
-    assertThat(output, contains(
-        Pair.of("data-audit", "media/audit.csv")
-    ));
+    scenario.mapSimpleValue("audit.csv", true);
 
     Path outputAudit = scenario.getOutputDir().resolve(scenario.getFormName() + " - audit.csv");
 
@@ -301,10 +298,7 @@ public class CsvFieldMappersTest {
     scenario = Scenario.nonGroup("some-form", DataType.BINARY, "audit", "meta");
     write(scenario.getWorkDir().resolve("audit.csv"), "event, node, start, end\nform start,,1536663986578,\n");
 
-    List<Pair<String, String>> output = scenario.mapSimpleValue("audit.csv", true);
-    assertThat(output, contains(
-        Pair.of("data-audit", "media/audit.csv")
-    ));
+    scenario.mapSimpleValue("audit.csv", true);
 
     Path outputAudit = scenario.getOutputDir().resolve(scenario.getFormName() + " - audit.csv");
 

